sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/subquery.scala

```diff
@@ -328,6 +328,8 @@ object PullupCorrelatedPredicates extends Rule[LogicalPlan] with PredicateHelper
     // Only a few unary nodes (Project/Filter/Aggregate) can contain subqueries.
     case q: UnaryNode =>
       rewriteSubQueries(q, q.children)
+    case s: SupportsSubquery =>
```

Comment:
Add a comment above this line? To be honest, it is hard to tell that this trait means UPDATE/MERGE/DELETE. 

Also, I think this change is just part of the whole changes for supporting the subquery in UPDATE/MERGE/DELETE. We need the other changes in Analyzer and Optimizer rules.  For example, CheckAnalysis.

Reply:
> Add a comment above this line? To be honest, it is hard to tell that this trait means UPDATE/MERGE/DELETE.

Sure, what kind of comment would make sense? `SupportsSubquery` seems generic to me and may cover different rules in the future. Here, I match the behavior in the analyzer.

> Also, I think this change is just part of the whole changes for supporting the subquery in UPDATE/MERGE/DELETE. We need the other changes in Analyzer and Optimizer rules. For example, CheckAnalysis.

You are right it is the first step and potentially more changes will be needed. At the same time, I think we've updated the analyzer already. Here is what we have in `CheckAnalysis`:

```
// Only certain operators are allowed to host subquery expression containing
// outer references.
plan match {
   case _: Filter | _: Aggregate | _: Project | _: SupportsSubquery => // Ok
   case other => failAnalysis(
       "Correlated scalar sub-queries can only be used in a " +
       s"Filter/Aggregate/Project and a few commands: $plan")
}
```

Reply:
If we decide to implement `SupportsSubquery` in other nodes and remove `UnaryNode` from here, I think the comment above may be sufficient (with minor tweaks once we remove `UnaryNode`).